torch/_functorch/_activation_checkpointing/ac_reorder_inference_pass.py

```diff
@@ -0,0 +1,158 @@
+            "Activation checkpoint reordering in fullgraph does not support RNG ops in checkpointed regions. "
+            "RNG state cannot be properly synchronized between "
+            "forward and recomputed backward operations. Please move RNG operations outside "
+            "of checkpoint regions, or use joint graph mode (where partitioner handles RNG)."
+        )
+
+    # Apply partitioner's edge case handling FIRST
+    # This marks nodes that must be saved (MUST_SAVE, collectives, mutations, etc.)
+    gm = cleanup_recompute_tags(gm, is_default_partition=False)
+
+    if not config.unsafe_allow_optimization_of_collectives:
+        force_save_collectives(gm)
+
+    force_save_bw_mutation_src(gm)
+
+    # Find backward boundary and build ordering
+    first_node_in_bwd = None
+    bwd_start = None
+    order = {}
+    for idx, node in enumerate(gm.graph.nodes):
+        order[node] = idx
+        if _is_backward_node(node) and first_node_in_bwd is None:
+            first_node_in_bwd = node
+            bwd_start = idx
+
+    if first_node_in_bwd is None:
+        return gm
+
+    # Build reordered graph
+    new_graph = fx.Graph()
+    env: dict[fx.Node, fx.Node] = {}
+    recomputed_nodes: dict[fx.Node, fx.Node] = {}
+
+    # Add placeholders
+    for node in gm.graph.find_nodes(op="placeholder"):
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    def gather_checkpointed_deps(node: fx.Node, visited: set) -> None:
+        if node in visited:
+            return
+        visited.add(node)
+        for inp in node.all_input_nodes:
+            if inp.op != "placeholder" and must_recompute(inp):
+                gather_checkpointed_deps(inp, visited)
+
+    # Insert forward nodes
+    for node in list(gm.graph.nodes)[:bwd_start]:
+        if node.op == "placeholder":
+            continue
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    # Insert backward nodes
+    for node in list(gm.graph.nodes)[bwd_start:]:
+        if node.op == "output":
+            continue
+
+        # Gather ALL checkpointed deps needed by this node
+        deps = set()
+        for inp in node.all_input_nodes:
+            if must_recompute(inp):
+                gather_checkpointed_deps(inp, deps)
+
+        # Insert deps in forward order, skipping already-inserted ones
+        for dep in sorted(deps, key=lambda n: order[n]):
+            if dep not in recomputed_nodes:
+
+                def get_input(x):
+                    if not isinstance(x, fx.Node):
+                        return x
+                    return recomputed_nodes.get(x, env[x])
+
+                dup = new_graph.node_copy(dep, get_input)
+                dup.name = dep.name + "_recomputed"
+                recomputed_nodes[dep] = dup
+
+        # Insert the backward node
+        def remat_input(x):
+            if not isinstance(x, fx.Node):
+                return x
+            return recomputed_nodes.get(x, env[x])
```

Comment:
This helper is the same as get_input, maybe we can just define it at the top level after env and recomputed_nodes are defined